gstreamer-source.c

```diff
@@ -316,6 +356,88 @@ const char *gstreamer_source_get_name(void *type_data)
 	return "GStreamer Source";
 }
 
+enum obs_media_state gstreamer_source_get_state(void *user_data)
+{
+	data_t *data = user_data;
+
+	return data->media_state;
+}
+
+int64_t gstreamer_source_get_time(void *user_data)
+{
+	data_t *data = user_data;
+	int64_t position;
+
+	if (gst_element_query_position(data->pipe, GST_FORMAT_TIME, &position))
+		if (GST_CLOCK_TIME_IS_VALID(position))
+			return GST_TIME_AS_MSECONDS(position);
+
+	return 0;
+}
+
+int64_t gstreamer_source_get_duration(void *user_data)
+{
+	data_t *data = user_data;
+	int64_t duration;
+
+	if (gst_element_query_duration(data->pipe, GST_FORMAT_TIME, &duration))
+		if (GST_CLOCK_TIME_IS_VALID(duration))
+			return GST_TIME_AS_MSECONDS(duration);
+
+	return 0;
+}
+
+void gstreamer_source_play_pause(void *user_data, bool pause)
+{
+	data_t *data = user_data;
+
+	gst_element_set_state(data->pipe,
+			      pause ? GST_STATE_PAUSED : GST_STATE_PLAYING);
+}
+
+void gstreamer_source_stop(void *user_data)
+{
+	data_t *data = user_data;
+
+	gst_element_set_state(data->pipe, GST_STATE_READY);
+}
+
+void gstreamer_source_restart(void *user_data)
+{
+	data_t *data = user_data;
+
+	gst_element_set_state(data->pipe, GST_STATE_PLAYING);
```

Comment:
I'm not sure what is expected for the pipeline. Is this a NOP when it is playing or should it seek back to the start and start over?

If it is supposed to start over probably something similar has to be done as in the bus callback EOS case and the restart flag being set. E.g. stoping the pipeline and rebuild it. (Note that simply changing states seems not enough as that breaks decodebin and friends)

Reply:
Yes, you're right -- this only appeared to be "working" in my initial testing because of what I (now) think is another bug!...

[ other bug (?) ]   When a bus error message occurs, the fallthrough to the EOS case from the ERROR case *always* sets the state to GST_STATE_NULL immediately -- before even checking the restart_on_ flags at all.  Should it really do that?  Quick smoke testing suggests this might be why my rtmp cameras (but not my rtsp cameras) usually flail quite a bit at first startup, requiring multiple restart attempts before finally getting lucky and starting properly.

I will experiment further with "[ other bug ]".

And I'll take another look at making `_restart` actually perform a full pipeline restart.

Reply:
I think when an error occurs it is usually an indication that the pipeline cannot continue. So I _believe_ it is okay to stop.

But if you see that in your RTMP cameras it may be interesting to know what these errors are.

The rebuilding of the pipeline is usually necessary for any elements that have "sometimes pad" like decodebin. The parse_launch methods will install signal handlers for connecting these pads when they become available. When you NULL->PLAY these again these handlers are no longer effective causing the pipeline to not connect correctly.

So maybe that is what you are seeing?

Reply:
Ok, turns out that my rtmp-camera-flailing problem was not caused by that GST_STATE_NULL state -- but it also turns out that I would *still* like to defer that state change anyway, to allow for "rewind" even after reaching EOS.  That is implemented in the first patch of my [v3] below.

Reply:
[v3 notes]

This is working great now for all OBS control and state-query operations, including detection of pipeline buffering.  All OBS media controls work as they should and (corrected from [v2]) the "restart" operation now completely destroys and recreates the gstreamer pipe.
 
Fwiw, the OBS gui does not actually show the buffering state in any way, but by adding extra logging to obs-gstreamer (not included here) I can verify that OBS websocket-query state does correctly indicate OBS_MEDIA_STATE_BUFFERING as long as the gstreamer pipeline is buffering.

I'm ready for this to be merged if you are.